File: src/training/episodes.ts
import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { pathToFileURL } from 'node:url'
import { digestJson } from './digest.js'
import { matchesHarnessRef } from './harness.js'
import { datasetDestination } from './snapshots.js'
import { parseContentRef, parseFeedbackRecord, parseGenerationReceipt, parsePolicyLease, parseTrainingEpisode, parseTrainingExternalBinding, parseTrainingRequest, requireContract, TrainingContractError } from './schema.js'
import { jsonProcess } from './process.js'
import { atomicWrite, TrainingContentStore, withTrainingFileLock } from './store.js'
import { ModelNodeTransport, syncContentGraph } from './transport.js'
import type { ContentRef, GenerationReceipt, PolicyLease, TrainingExternalBinding, TrainingHandle, TrainingRequestV2 } from './types.js'
import type { TrainingDeploymentConfig } from './types.js'
import { assertHitchCommit, observeHitchController, verifyExecutionPlacement } from './placement-observation.js'

type Json = Record<string, unknown>
const object = (value: unknown): Json => { requireContract(!!value && typeof value === 'object' && !Array.isArray(value), 'invalid-episode-data', 'expected structured episode data'); return value as Json }
const text = (value: unknown): string => { requireContract(typeof value === 'string' && !!value, 'invalid-episode-data', 'expected a nonempty episode identity'); return value }
const list = (value: unknown): unknown[] => { requireContract(Array.isArray(value), 'invalid-episode-data', 'expected an episode array'); return value }
const same = (a: unknown, b: unknown): boolean => digestJson(a) === digestJson(b)

export interface RolloutIntent {
  schemaVersion: 2; id: string; jobId: string; incarnation: string; trainingRunId: string; inputDigest: string; key: string
  context: Json; binding: TrainingExternalBinding; lease: PolicyLease; gateway: { nodePort: number }; credential: string
}
interface IntentEntry { intent: RolloutIntent; cancelRequested: boolean; ack: { evalId: string } | null; result: Json | null }
interface ControllerEpisodeState { intentDigest: string; submitted: boolean; evalId?: string; result?: Json }
export interface TrainingEpisodeControllerOptions {
  command: string[]; root: string; workspace: string; harnessSourceDirectory: string; python: string[]; episodeTimeoutSeconds: number
  deployment?: TrainingDeploymentConfig
}

export function episodeAddress(intent: RolloutIntent) {
  return { id: intent.id, inputDigest: intent.inputDigest, jobId: intent.jobId, incarnation: intent.incarnation,
    batchId: intent.lease.batchId, policyVersion: intent.lease.policyVersion, fencingToken: intent.lease.fencingToken }
}

/** Validate against the train-only request, never against fields supplied by the task agent. */
export function validateRolloutIntent(value: unknown, request: TrainingRequestV2, handle: TrainingHandle): RolloutIntent {
  const input = object(value); const { inputDigest, ...body } = input
  requireContract(input.schemaVersion === 2 && inputDigest === digestJson(body) && input.jobId === handle.jobId
    && input.trainingRunId === request.trainingRunId && typeof input.incarnation === 'string', 'episode-intent-drift', 'node intent changed its job, request or input digest')
  const context = object(input.context), lease = parsePolicyLease(input.lease), binding = parseTrainingExternalBinding(input.binding)
  const task = request.trainDataset.tasks.find(task => task.id === context.taskId)
  const sampling = request.rollout.sampling
  requireContract(task && same(context.taskRef, task.taskRef) && same(context.environmentRef, task.environmentRef)
    && same(context.harnessRef, request.fixedHarness.manifestRef) && context.id === input.id && context.trainingRunId === request.trainingRunId
    && context.runId === null && context.logicalAttempt === 1 && Number.isSafeInteger(context.slot) && Number(context.slot) >= 0 && Number(context.slot) < request.rollout.groupSize
    && context.policyVersion === lease.policyVersion && context.runtimeInstanceId === lease.runtimeInstanceId && context.batchId === lease.batchId
    && context.wireModel === lease.policyVersion && context.tokenizerDigest === request.parentModel.tokenizerDigest && context.chatTemplateDigest === request.parentModel.chatTemplateDigest
    && context.generationContractDigest === request.trainer.runtimeLock.protocolDigest && context.verifierVersion === request.verifier.digest
    && context.maxContextTokens === sampling.maxContextTokens && context.maxEpisodeSteps === request.budgets.maxEpisodeSteps && context.maxRolloutTokens === request.budgets.maxRolloutTokens
    && same(context.sampling, { temperature: 1, top_p: 1, top_k: -1, repetition_penalty: 1, max_new_tokens: sampling.maxNewTokens,
      skip_special_tokens: false, spaces_between_special_tokens: false, no_stop_trim: true }),
  'episode-assignment-drift', 'rollout intent is not an authorized frozen train slot')
  requireContract(lease.trainingRunId === request.trainingRunId && lease.parentModelVersionId === request.parentModel.id && lease.state === 'serving'
    && binding.trainingRunId === request.trainingRunId && binding.bindingId === context.bindingId && binding.expectedPolicyVersion === lease.policyVersion
    && binding.fencingToken === lease.fencingToken && binding.expiresAt === lease.expiresAt && binding.policyLeaseRef.digest === digestJson(lease)
    && binding.generationContractDigest === request.trainer.runtimeLock.protocolDigest && binding.maxOutputTokens === sampling.maxNewTokens
    && binding.maxEpisodeSteps === request.budgets.maxEpisodeSteps && lease.samplingDigest === digestJson(context.sampling),
  'episode-policy-drift', 'episode binding changed its policy, fence or generation contract')
  requireContract(input.key === digestJson([request.trainingRunId, lease.batchId, task.taskRef.digest, context.groupId, context.slot])
    && /^[a-f0-9]{64}$/.test(text(input.credential)) && Number.isInteger(object(input.gateway).nodePort), 'episode-submission-drift', 'episode key, credential or route is invalid')
  return { ...input, context, lease, binding } as unknown as RolloutIntent
}

/** Controller-owned journals reconcile public Hitch operations one logical slot at a time. */
export class TrainingEpisodeCoordinator {
  constructor(readonly store: TrainingContentStore, readonly transport: ModelNodeTransport, readonly options: TrainingEpisodeControllerOptions,
    private readonly invoke = jsonProcess, private readonly fetcher: typeof fetch = fetch) {}
  private path(jobId: string): string {
    requireContract(/^job_[a-f0-9]{32}$/.test(jobId), 'invalid-job-handle', 'controller journal requires a canonical node job ID')
    return join(this.options.workspace, 'training-episodes', jobId)
  }
  private async call(args: string[], payload?: unknown, timeout = 60_000): Promise<unknown> {
    try { return await this.invoke(this.options.command, ['--root', this.options.root, ...args], payload, timeout) }
    catch (error) {
      if (error instanceof TrainingContractError) error.message = `Hitch ${args.slice(0, 2).join(' ')}: ${error.message}`
      throw error
    }
  }
  async preflight(request: TrainingRequestV2): Promise<void> {
    if (this.options.deployment) await verifyExecutionPlacement(request.deployment, this.options.deployment, this.options, request.trainer.runtimeLock.hitchCommit, this.invoke)
    else assertHitchCommit(await observeHitchController(this.options, this.invoke), request.trainer.runtimeLock.hitchCommit)
    const caps = object(await this.call(['capabilities', '--json']))
    for (const capability of ['training_external_binding', 'exact_policy_tokens', 'training_policy_fencing']) {
      requireContract(caps[capability] === '1', 'hitch-capability-missing', `controller Hitch lacks ${capability}`)
    }
    requireContract(list(caps.training_harnesses).includes(request.fixedHarness.adapter), 'hitch-harness-missing', 'controller Hitch lacks the fixed training harness')
    if (request.deployment.taskExecution.placement === 'remote') requireContract(caps.remote_training_external_binding === '2', 'remote-training-capability-missing', 'selected provider requires the versioned remote training binding/capture capability')
    for (const task of request.trainDataset.tasks) {
      await this.store.readBytes(task.taskRef); await this.store.readBytes(task.environmentRef)
    }
    await this.store.readBytes(request.fixedHarness.manifestRef); await this.store.readBytes(request.verifier)
  }
  async remember(request: TrainingRequestV2, jobId: string): Promise<void> {
    const path = join(this.path(jobId), 'request.json')
    await withTrainingFileLock(path + '.lock', async () => {
      let old: unknown
      try { old = JSON.parse(await readFile(path, 'utf8')) } catch (error) { if ((error as NodeJS.ErrnoException).code !== 'ENOENT') throw error }
      requireContract(!old || same(old, request), 'episode-request-drift', 'controller job request changed')
      if (!old) await atomicWrite(path, request)
    })
  }
  private async entries(handle: TrainingHandle, renew: boolean): Promise<IntentEntry[]> {
    const entries: IntentEntry[] = []; let cursor = 0
    do {
      const response = object(await this.transport.call('training.episodes.list', { handle, renew: renew && cursor === 0, cursor }))
      requireContract(response.schemaVersion === 2 && response.jobId === handle.jobId, 'episode-job-drift', 'node journal belongs to another job')
      entries.push(...list(response.entries) as unknown as IntentEntry[])
      if (response.nextCursor === null) return entries
      requireContract(Number.isSafeInteger(response.nextCursor) && Number(response.nextCursor) > cursor, 'episode-cursor-drift', 'node journal cursor must advance')
      cursor = Number(response.nextCursor)
    } while (true)
  }
  async reconcile(handle: TrainingHandle, cancel = false): Promise<{ pending: boolean }> {
    return this.withControl(handle, reconcile => reconcile(cancel))
  }
  async withControl<R>(handle: TrainingHandle, action: (reconcile: (cancel?: boolean) => Promise<{ pending: boolean }>) => Promise<R>): Promise<R> {
    return withTrainingFileLock(join(this.path(handle.jobId), 'coordinator.lock'), () => action(cancel => this.reconcileLocked(handle, cancel)))
  }
  private async reconcileLocked(handle: TrainingHandle, cancel = false): Promise<{ pending: boolean }> {
    const request = parseTrainingRequest(JSON.parse(await readFile(join(this.path(handle.jobId), 'request.json'), 'utf8')))
    requireContract(request.schemaVersion === 2 && digestJson(request) === handle.requestDigest, 'episode-request-drift', 'controller journal is not bound to this node job')
    let heartbeat: Promise<void> | undefined; let contactError: unknown
    const fenced = new Set<string>()
    const observe = (entries: IntentEntry[]) => { for (const entry of entries) if (entry.cancelRequested) fenced.add(entry.intent.id) }
    const timer = cancel ? undefined : setInterval(() => {
      if (heartbeat) return
      heartbeat = this.entries(handle, true).then(observe).catch(error => { contactError = error }).finally(() => { heartbeat = undefined })
    }, 2_000)
    try {
      const entries = await this.entries(handle, !cancel)
      observe(entries)
      if (!cancel && this.options.deployment) await verifyExecutionPlacement(request.deployment, this.options.deployment, this.options, request.trainer.runtimeLock.hitchCommit, this.invoke)
      for (const entry of entries) {
        const intent = validateRolloutIntent(entry.intent, request, handle)
        const serving = () => {
          if (contactError || fenced.has(intent.id)) throw new TrainingContractError('controller-contact-lost', 'model-node contact or policy lease was lost; stop dispatching new episodes')
        }
        await this.reconcileOne(handle, request, { ...entry, intent }, cancel || entry.cancelRequested, serving)
      }
      return { pending: (await this.entries(handle, false)).some(entry => !entry.result) }
    } finally { if (timer) clearInterval(timer); await heartbeat }
  }
  private async reconcileOne(handle: TrainingHandle, request: TrainingRequestV2, entry: IntentEntry, cancel: boolean, serving: () => void): Promise<void> {
    const { intent } = entry; const address = episodeAddress(intent)
    const directory = join(this.path(handle.jobId), digestJson(intent.id).slice(7)), statePath = join(directory, 'state.json')
    let state: ControllerEpisodeState
    try { state = JSON.parse(await readFile(statePath, 'utf8')) } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== 'ENOENT') throw error
      state = { intentDigest: intent.inputDigest, submitted: false }; await atomicWrite(statePath, state)
    }
    requireContract(state.intentDigest === intent.inputDigest, 'episode-intent-drift', 'controller intent journal changed')
    await atomicWrite(join(directory, 'intent.json'), intent)
    if (entry.result) {
      requireContract(state.result && same(entry.result, state.result), 'episode-result-drift', 'node result differs from durable controller feedback')
      return
    }
    const resolve = async (result: Json) => {
      state.result = result; await atomicWrite(statePath, state)
      await this.transport.call('training.episodes.result', { handle, address, result })
    }
    if (state.result) { await resolve(state.result); return }
    if (entry.ack) {
      requireContract(!state.evalId || state.evalId === entry.ack.evalId, 'episode-ack-conflict', 'controller and node disagree about canonical eval')
      state.evalId = entry.ack.evalId; state.submitted = true; await atomicWrite(statePath, state)
    }
    if (cancel && !state.submitted) { await resolve({ schemaVersion: 2, outcome: 'cancelled', evalId: null, reason: 'cancelled' }); return }
    const harness = `${request.fixedHarness.adapter}@git+${pathToFileURL(this.options.harnessSourceDirectory).href}#${request.fixedHarness.commit}`
    const task = request.trainDataset.tasks.find(task => task.id === intent.context.taskId)!
    const datasetRoot = join(this.options.workspace, 'datasets', task.taskRef.digest.slice(7))
    const dataset = datasetDestination(await this.store.readJson(task.taskRef), datasetRoot)
    if (!state.evalId) {
      if (!state.submitted) {
        serving()
        await this.invoke(this.options.python, ['-m', 'gear_training.artifacts', 'materialize', '--store-root', this.store.root], { ref: task.taskRef, destination: datasetRoot }, 3_600_000)
        await this.transport.download(this.store, intent.binding.policyLeaseRef)
        const endpoint = await this.transport.gateway(this.options.workspace, intent.gateway.nodePort)
        const response = await this.fetcher(`${endpoint}/v1/lease`, { headers: { Authorization: `Bearer ${intent.credential}` }, signal: AbortSignal.timeout(10_000) })
        requireContract(response.ok, 'training-route-unavailable', 'controller cannot reach the exact generation lease')
        const lease = object(await response.json())
        requireContract(lease.episodeId === intent.id && lease.trainingRunId === request.trainingRunId && lease.policyVersion === intent.lease.policyVersion
          && lease.fencingToken === intent.lease.fencingToken && lease.state === 'serving' && lease.capture === 'exact-policy-tokens-v1'
          && lease.generationContractDigest === request.trainer.runtimeLock.protocolDigest, 'training-route-drift', 'gateway route points to another policy or episode')
        serving()
        const registered = object(await this.call(['training', 'register', '--file', '-'], { schema_version: '1', binding: intent.binding, base_url: endpoint + '/v1', credential: intent.credential }))
        requireContract(same(registered.binding, intent.binding), 'training-binding-drift', 'Hitch registration changed the frozen binding')
        await atomicWrite(join(directory, 'binding.json'), intent.binding)
        serving(); state.submitted = true; await atomicWrite(statePath, state)
      }
      // Replay the exact durable submission after a lost reply, including during
      // cancellation. Its old lease cannot be relabeled as a fresh generation.
      const accepted = object(await this.call(['eval', 'submit', '--dataset', dataset, '--harness', harness, '--model', `training/${intent.binding.bindingId}`,
        '--attempts', '1', '--max-concurrent', '1', '--infrastructure-retries', '0', '--training-binding-file', join(directory, 'binding.json'),
        '--provider', request.deployment.taskExecution.provider, '--model-capture', 'proxy', '--require-model-capture',
        '--timeout', `${this.options.episodeTimeoutSeconds}s`, '--idempotency-key', intent.key]))
      state.evalId = text(accepted.eval_id); await atomicWrite(statePath, state)
    }
    await this.transport.call('training.episodes.ack', { handle, address, ack: { evalId: state.evalId } })
    if (cancel) await this.call(['eval', 'cancel', state.evalId])
    const inspection = object(await this.call(['eval', 'inspect', state.evalId, '--json']))
    const submission = object(inspection.request), acceptedRequest = object(submission.request ?? submission)
    requireContract(same(acceptedRequest.training_binding, intent.binding), 'hitch-binding-drift', 'Hitch inspection lost the episode binding')
    requireContract(acceptedRequest.harness_ref === harness, 'hitch-harness-source-drift', 'accepted Hitch submission changed the frozen harness source')
    if (!inspection.result || ['running', 'cancelling'].includes(String(object(inspection.control ?? {}).state))) return
    const result = object(inspection.result), trials = list(result.trials)
    requireContract(['succeeded', 'failed', 'cancelled', 'timed_out'].includes(String(result.status)), 'hitch-terminal-drift', 'unexpected Hitch terminal status')
    if (cancel) { await resolve({ schemaVersion: 2, outcome: 'cancelled', evalId: state.evalId, reason: 'cancelled' }); return }
    if (trials.length !== 1 || typeof object(trials[0]).run_id !== 'string') {
      await resolve({ schemaVersion: 2, outcome: 'rejected', evalId: state.evalId, reason: 'missing-canonical-run' }); return
    }
    const trial = object(trials[0]), runId = text(trial.run_id)
    const native = object(await this.transport.call('training.episodes.receipts', { handle, address }))
    requireContract(native.runId === runId, 'hitch-canonical-run-mismatch', 'verifier and generation used different physical runs')
    if (!native.complete) { await resolve({ schemaVersion: 2, outcome: 'rejected', evalId: state.evalId, reason: 'incomplete-receipts' }); return }
    const receiptRefs = list(native.receiptRefs).map(parseContentRef)
    await syncContentGraph(this.transport, this.store, receiptRefs, 'download')
    const receipts: GenerationReceipt[] = []
    for (const ref of receiptRefs) receipts.push(parseGenerationReceipt(await this.store.readJson(ref)))
    const loaded = object(await this.call(['runs', 'inspect', runId, '--json']))
    const verifier = object(await this.call(['verifier', 'inspect', runId, '--json']))
    const terminal = object(await this.call(['training', 'evidence', runId, '--json']))
    const record = object(loaded.record), context = object(record.context), parent = object(record.parent), model = object(record.model), protocol = object(record.protocol), actualHarness = object(record.harness)
    const environment = object(await this.store.readJson(task.environmentRef)), expectedHarness = object(object(await this.store.readJson(request.fixedHarness.manifestRef)).hitch)
    requireContract(loaded.record_status === 'valid' && loaded.trajectory_status !== 'corrupt' && record.run_id === runId
      && parent.eval_id === state.evalId && parent.attempt === 1 && (trial.attempt ?? 1) === 1 && trial.task_id === task.id
      && context.kind === 'benchmark_task' && context.task_id === task.id && context.task_digest === environment.taskDigest && context.verifier_identity === environment.verifierIdentity
      && protocol.environment_identity === environment.hitchEnvironmentIdentity && model.provider === 'slime-training' && model.effective_id === intent.lease.policyVersion && model.identity_resolved === true
      && actualHarness.harness_id === expectedHarness.harnessId && actualHarness.revision_identity === expectedHarness.revisionIdentity && actualHarness.artifact_id === expectedHarness.artifactId
      && matchesHarnessRef(actualHarness.requested_ref, harness, request.fixedHarness.adapter, request.fixedHarness.commit)
      && object(terminal.training_external).policy_version === intent.lease.policyVersion,
    'training-canonical-identity-drift', 'canonical task/verifier/environment/harness/policy differs from its frozen training slot')
    const evidenceRef = await this.store.putJson({ inspection, loaded, verifier, terminal })
    const observation = object(record.observation ?? {})
    if (trial.observation_status !== 'valid' || observation.status !== 'valid' || observation.reward !== trial.reward
      || object(verifier.verifier).status !== 'complete' || terminal.termination !== 'terminated' || !Number.isFinite(trial.reward)) {
      await resolve({ schemaVersion: 2, outcome: 'rejected', evalId: state.evalId, reason: 'invalid-canonical-observation', evidenceRef }); return
    }
    const verificationRef = await this.store.putJson({ schemaVersion: 2, kind: 'controller-verifier-observation', runId, verifierVersion: request.verifier.digest,
      reward: trial.reward, valid: true, sourceEvidenceRef: evidenceRef })
    const feedback = parseFeedbackRecord({ schemaVersion: 1, id: `feedback_${digestJson([intent.id, evidenceRef]).slice(7, 39)}`, episodeId: intent.id, runId,
      receiptIds: receipts.map(receipt => receipt.id), verifierVersion: request.verifier.digest, verifierEvidenceRef: verificationRef, outcome: 'valid', reward: trial.reward })
    const episode = parseTrainingEpisode({ schemaVersion: 1, id: intent.id, groupId: intent.context.groupId, slot: intent.context.slot, harnessRef: request.fixedHarness.manifestRef,
      taskRef: task.taskRef, environmentRef: task.environmentRef, policyVersion: intent.lease.policyVersion, runId, receiptIds: feedback.receiptIds, feedbackId: feedback.id,
      termination: 'terminated', eligibility: 'eligible', rejectionReasons: [] })
    const assembly = { schemaVersion: 2, kind: 'controller-episode-verification', episodeId: intent.id, runId, taskDigest: task.taskRef.digest,
      environmentDigest: task.environmentRef.digest, harnessDigest: request.fixedHarness.manifestRef.digest, verifierVersion: request.verifier.digest, feedbackDigest: digestJson(feedback) }
    const feedbackRef = await this.store.putJson(feedback), episodeRef = await this.store.putJson(episode), assemblyRef = await this.store.putJson(assembly)
    // Only these projections travel to the trainer. Raw canonical/verifier data
    // and all task files remain in the controller CAS, referenced for collection.
    for (const ref of [verificationRef, feedbackRef, episodeRef, assemblyRef]) await this.transport.upload(this.store, ref)
    const resolution = { schemaVersion: 2, outcome: 'feedback', evalId: state.evalId, feedbackRef, episodeRef, assemblyRef }
    // Establish exact native-sample admission before journaling immutable
    // feedback. A lost reply can retry; a known invalid sample is rejected.
    const admission = object(await this.transport.call('training.episodes.admit', { handle, address, result: resolution }))
    requireContract(typeof admission.valid === 'boolean', 'invalid-episode-admission', 'node did not establish sample admission')
    if (!admission.valid) {
      const reason = text(admission.reason)
      requireContract(/^[a-z][a-z0-9-]{0,127}$/.test(reason), 'invalid-episode-admission', 'node admission needs a bounded reason')
      await resolve({ schemaVersion: 2, outcome: 'rejected', evalId: state.evalId, reason, evidenceRef }); return
    }
    await resolve(resolution)
  }
}

File: python/gear_training/episodes.py
"""V2 durable rollout intents and controller feedback, colocated with receipts.

Only authenticated node RPC can acknowledge or resolve an intent. Generation
credentials authorize the native gateway, never this control protocol.
"""
from __future__ import annotations

import json
import re
import time
from pathlib import Path

from .content import ContentStore, ContractError, canonical, digest_bytes, digest_json, require
from .ledger import Ledger
from .state import load, lock


class EpisodeJournal:
    def __init__(self, directory, ledger=None):
        self.directory = Path(directory)
        self.ledger = ledger or Ledger(self.directory / "ledger.sqlite")
        self.owned_connection = ledger is None
        self.request = load(self.directory / "request.json")
        self.config = load(self.directory / "config.json")
        require(self.request["schemaVersion"] == 2 and self.config["schemaVersion"] == 2, "episode-protocol-version", "controller episodes require a v2 job")
        self.store = ContentStore(self.config["storeRoot"])
        self.ledger.db.executescript("""
          CREATE TABLE IF NOT EXISTS episode_intents(id TEXT PRIMARY KEY, body TEXT NOT NULL, input_digest TEXT NOT NULL,
            ack TEXT, result TEXT, consumed INTEGER NOT NULL DEFAULT 0, cancel_requested INTEGER NOT NULL DEFAULT 0);
        """)

    def close(self):
        if self.owned_connection: self.ledger.close()

    def publish(self, context, binding, credential, node_port):
        worker = load(self.directory / "worker.json")
        lease = self.ledger.assert_serving(context["batchId"], context["runtimeInstanceId"], binding["fencingToken"])
        registered = self.ledger.db.execute("SELECT token_hash FROM episodes WHERE id=?", (context["id"],)).fetchone()
        require(registered and registered[0] == digest_bytes(credential.encode()), "episode-credential-drift", "intent credential does not belong to this exact episode")
        require(self.store.read_json(binding["policyLeaseRef"]) == lease, "episode-lease-drift", "binding does not identify this policy lease")
        body = {"schemaVersion": 2, "id": context["id"], "jobId": self.directory.name, "incarnation": worker["incarnation"],
                "trainingRunId": self.request["trainingRunId"], "context": context, "binding": binding, "lease": lease,
                "key": digest_json([self.request["trainingRunId"], lease["batchId"], context["taskRef"]["digest"], context["groupId"], context["slot"]]),
                "gateway": {"nodePort": node_port}, "credential": credential}
        encoded, digest = canonical(body), digest_json(body)
        with self.ledger.transaction():
            old = self.ledger.db.execute("SELECT body FROM episode_intents WHERE id=?", (context["id"],)).fetchone()
            require(not old or old[0] == encoded, "episode-intent-conflict", "logical episode intent cannot change")
            self.ledger.db.execute("INSERT OR IGNORE INTO episode_intents(id,body,input_digest) VALUES(?,?,?)", (context["id"], encoded, digest))
        return {**body, "inputDigest": digest}

    def _row(self, address):
        row = self.ledger.db.execute("SELECT * FROM episode_intents WHERE id=?", (address.get("id"),)).fetchone()
        require(row, "unknown-episode-intent", "episode intent does not exist")
        intent = json.loads(row["body"])
        expected = {"id": intent["id"], "inputDigest": row["input_digest"], "jobId": intent["jobId"], "incarnation": intent["incarnation"],
                    "batchId": intent["lease"]["batchId"], "policyVersion": intent["lease"]["policyVersion"], "fencingToken": intent["lease"]["fencingToken"]}
        require(address == expected, "episode-address-drift", "episode control address differs from the job, incarnation or policy fence")
        return row, intent

    def list(self, *, renew=False, cursor=0):
        require(type(cursor) is int and cursor >= 0, "invalid-episode-cursor", "episode cursor must be a nonnegative row ID")
        if renew and not (self.directory / "cancel.json").exists():
            for row in self.ledger.db.execute("SELECT batch FROM controller_contacts").fetchall():
                if self.ledger.lease(row[0])["state"] == "serving": self.ledger.renew_controller(row[0], self.config["controllerTimeoutSeconds"])
        entries = []
        rows = self.ledger.db.execute("SELECT rowid AS cursor,* FROM episode_intents WHERE consumed=0 AND rowid>? ORDER BY rowid LIMIT 65", (cursor,)).fetchall()
        for row in rows[:64]:
            body = json.loads(row["body"]); lease = self.ledger.lease(body["lease"]["batchId"])
            contact = self.ledger.db.execute("SELECT expires FROM controller_contacts WHERE batch=?", (lease["batchId"],)).fetchone()
            cancelled = bool(row["cancel_requested"] or (self.directory / "cancel.json").exists() or lease["state"] != "serving"
                             or (contact and time.time() >= contact[0]))
            entries.append({"intent": {**body, "inputDigest": row["input_digest"]}, "cancelRequested": cancelled,
                            "ack": json.loads(row["ack"]) if row["ack"] else None, "result": json.loads(row["result"]) if row["result"] else None})
        return {"schemaVersion": 2, "jobId": self.directory.name, "entries": entries,
                "nextCursor": rows[63]["cursor"] if len(rows) > 64 else None}

    def acknowledge(self, address, ack):
        require(isinstance(ack, dict) and set(ack) == {"evalId"} and re.fullmatch(r"eval_[a-f0-9]{32}", str(ack["evalId"])), "invalid-episode-ack", "acknowledgement needs a canonical eval ID")
        with self.ledger.transaction():
            row, _ = self._row(address)
            require(not row["ack"] or row["ack"] == canonical(ack), "episode-ack-conflict", "episode acknowledgement cannot change its eval")
            require(not (row["consumed"] or row["result"]) or row["ack"], "episode-already-cancelled", "cannot submit an undispatched cancelled slot")
            self.ledger.db.execute("UPDATE episode_intents SET ack=? WHERE id=?", (canonical(ack), address["id"]))
        return {"ack": ack}

    def receipts(self, address, *, confirmed_stopped=False):
        _, intent = self._row(address)
        if confirmed_stopped:
            for row in self.ledger.db.execute("SELECT id FROM requests WHERE episode=? AND state='pending'", (intent["id"],)).fetchall():
                self.ledger.fail_request(row[0], confirmed_stopped=True)
        # Canonical terminal/cancellation is established by the controller before
        # it asks for closure. Native generation/receipt writes must also finish.
        self.ledger.finish_episode(intent["id"])
        rows = self.ledger.db.execute("SELECT state,receipt FROM requests WHERE episode=? ORDER BY call_index", (intent["id"],)).fetchall()
        return {"runId": self.ledger.context(intent["id"])["runId"], "complete": bool(rows) and all(r[0] == "complete" for r in rows),
                "receiptRefs": [json.loads(r[1]) for r in rows if r[1]]}

    def admit(self, address, result, *, confirmed_stopped=False):
        row, intent = self._row(address)
        require(result.get("schemaVersion") == 2 and result.get("outcome") == "feedback", "invalid-episode-result", "admission requires feedback")
        ack = json.loads(row["ack"]) if row["ack"] else None
        require(ack and result.get("evalId") == ack["evalId"], "episode-eval-drift", "feedback refers to another eval")
        receipts = self.receipts(address, confirmed_stopped=confirmed_stopped)
        require(receipts["complete"], "incomplete-episode", "feedback requires complete native receipts")
        feedback, episode, assembly = (self.store.read_json(result[key]) for key in ("feedbackRef", "episodeRef", "assemblyRef"))
        actual = self.ledger.context(intent["id"])
        require(feedback["runId"] == episode["runId"] == receipts["runId"] and episode["id"] == intent["id"], "episode-run-drift", "feedback and generation refer to different canonical runs")
        require(feedback["receiptIds"] == [self.store.read_json(ref)["id"] for ref in receipts["receiptRefs"]], "episode-receipt-drift", "feedback ordered receipts differ from native generation")
        from .samples import build_episode
        try:
            build_episode(self.store, episode, [self.store.read_json(ref) for ref in receipts["receiptRefs"]], feedback, actual, assembly=assembly)
        except ContractError as error:
            # Only exact-sample admission is a bounded group rejection. Missing
            # or corrupt CAS objects and control identity conflicts stay errors.
            if error.code in ("corrupt-content", "content-digest-mismatch", "invalid-ref", "invalid-digest", "invalid-content-uri", "invalid-media-type", "non-finite-json"): raise
            return {"valid": False, "reason": error.code}
        return {"valid": True}

    def resolve(self, address, result, *, confirmed_stopped=False):
        # Serialize control mutations across RPC processes. The driver consumes
        # only committed results, so a crash after add_feedback is replayable.
        with lock(self.directory / "episode-resolution.lock"):
            return self._resolve(address, result, confirmed_stopped=confirmed_stopped)

    def _resolve(self, address, result, *, confirmed_stopped=False):
        row, intent = self._row(address)
        if row["result"]:
            require(row["result"] == canonical(result), "episode-result-conflict", "episode feedback cannot change on retry")
            return {"accepted": True}
        require(isinstance(result, dict) and result.get("schemaVersion") == 2 and result.get("outcome") in ("feedback", "cancelled", "rejected"),
                "invalid-episode-result", "unknown episode resolution")
        ack = json.loads(row["ack"]) if row["ack"] else None
        require(result.get("evalId") == (ack["evalId"] if ack else None), "episode-eval-drift", "feedback refers to another eval")
        receipts = self.receipts(address, confirmed_stopped=confirmed_stopped)
        if result["outcome"] == "feedback":
            admission = self.admit(address, result, confirmed_stopped=confirmed_stopped)
            require(admission["valid"], admission.get("reason", "invalid-episode"), "exact sample failed admission")
            self.ledger.add_feedback(self.store.read_json(result["feedbackRef"]))
        else:
            require(isinstance(result.get("reason"), str) and re.fullmatch(r"[a-z][a-z0-9-]{0,127}", result["reason"]), "invalid-episode-rejection", "rejection needs a bounded reason code")
            require(ack or (receipts["runId"] is None and not receipts["receiptRefs"]), "undispatched-episode-generated", "an undispatched cancellation cannot hide native generation")
        with self.ledger.transaction():
            row, _ = self._row(address)
            require(not row["result"] or row["result"] == canonical(result), "episode-result-conflict", "concurrent feedback changed")
            self.ledger.db.execute("UPDATE episode_intents SET result=? WHERE id=?", (canonical(result), intent["id"]))
        return {"accepted": True}

    def cancel(self, intent_id):
        self.ledger.db.execute("UPDATE episode_intents SET cancel_requested=1 WHERE id=?", (intent_id,))

    def result(self, intent_id):
        row = self.ledger.db.execute("SELECT result FROM episode_intents WHERE id=?", (intent_id,)).fetchone()
        return json.loads(row[0]) if row and row[0] else None

    def consume(self, intent_id):
        with self.ledger.transaction():
            row = self.ledger.db.execute("SELECT result FROM episode_intents WHERE id=?", (intent_id,)).fetchone()
            require(row and row[0], "episode-result-pending", "cannot consume an unresolved logical slot")
            self.ledger.db.execute("UPDATE episode_intents SET consumed=1 WHERE id=?", (intent_id,))

    def reconcile_stopped(self):
        pending = self.ledger.db.execute("SELECT id FROM episode_intents WHERE result IS NULL").fetchone()
        require(not pending, "controller-episodes-pending", "controller must reconcile old Hitch slots before a new trainer incarnation")
        for row in self.ledger.db.execute("SELECT id FROM requests WHERE state='pending'").fetchall(): self.ledger.fail_request(row[0], confirmed_stopped=True)
        for row in self.ledger.db.execute("SELECT id FROM episodes WHERE state='running'").fetchall(): self.ledger.finish_episode(row[0])
        for row in self.ledger.db.execute("SELECT batch FROM leases WHERE state!='closed'").fetchall():
            self.ledger.drain(row[0]); self.ledger.close_lease(row[0])
        self.ledger.db.execute("UPDATE episode_intents SET consumed=1 WHERE result IS NOT NULL")
